storage/replica.go

```diff
@@ -542,37 +608,106 @@ func (r *Replica) checkCmdHeader(header *proto.RequestHeader) error {
-// commands which overlap its key range. This method will block if
-// there are any overlapping commands already in the queue. Returns
-// the command queue insertion key, to be supplied to subsequent
-// invocation of endCmd().
-func (r *Replica) beginCmd(header *proto.RequestHeader, readOnly bool) interface{} {
-	r.Lock()
-	var wg sync.WaitGroup
-	r.cmdQ.GetWait(header.Key, header.EndKey, readOnly, &wg)
-	cmdKey := r.cmdQ.Add(header.Key, header.EndKey, readOnly)
-	r.Unlock()
-	wg.Wait()
+// checkBatchRequest verifies BatchRequest validity requirements. In
+// particular, timestamp, user, user priority and transactions must
+// all be set to identical values between the batch request header and
+// all constituent batch requests. Also, either all requests must be
+// read-only, or none.
+func (r *Replica) checkBatchRequest(bArgs *proto.BatchRequest) error {
+	var isReadOnly bool
+	for i := range bArgs.Requests {
+		args := bArgs.Requests[i].GetValue().(proto.Request)
+		header := args.Header()
+		if args.Method() == proto.EndTransaction && len(bArgs.Requests) != 1 {
+			return util.Errorf("cannot mix EndTransaction with other operations in a batch")
+		}
+		// Compare only WallTime since logical ticks are used for self-overlapping
+		// batches (see setBatchTimestamps()).
+		if header.Timestamp.WallTime != bArgs.Timestamp.WallTime {
+			return util.Errorf("conflicting timestamp %s on call in batch at %s", header.Timestamp, bArgs.Timestamp)
+		}
+		if header.GetUserPriority() != bArgs.GetUserPriority() {
+			return util.Errorf("conflicting user priority on call in batch")
+		}
+		if !header.Txn.Equal(bArgs.Txn) {
+			return util.Errorf("conflicting transaction on call in transactional batch")
+		}
+		if proto.IsReadOnly(args) && header.ReadConsistency == proto.INCONSISTENT {
+			// Disallow any inconsistent reads within txns.
+			if header.Txn != nil {
+				return util.Errorf("cannot allow inconsistent reads within a transaction")
+			}
+		} else if proto.IsReadOnly(args) && header.ReadConsistency == proto.CONSENSUS {
+			return util.Errorf("consensus reads not implemented")
+		}
+		if i == 0 {
+			isReadOnly = proto.IsReadOnly(args)
+		} else if proto.IsReadOnly(args) != isReadOnly {
+			return util.Errorf("batch mixes read-only and write requests")
+		}
+	}
+	return nil
+}
+
+// beginCmds waits for any overlapping, already-executing commands via
+// the command queue and adds itself to queues based on keys affected by the
+// batched commands. This gates subsequent commands with overlapping keys or
+// key ranges. This method will block if there are any overlapping commands
+// already in the queue. Returns the command queue insertion keys, to be
+// supplied to a subsequent invocation of endCmds().
+func (r *Replica) beginCmds(bArgs *proto.BatchRequest) ([]interface{}, error) {
+	var cmdKeys []interface{}
+	// Don't use the command queue for inconsistent reads.
+	if bArgs.ReadConsistency != proto.INCONSISTENT {
```

Comment:
Do we also need to check `IsReadOnly` here? `checkBatchRequest` validates the `ReadConsistency` field for read-only requests, but appears to let anything through for non-read-only requests.